src/api/erc20/Erc20ApiImpl.ts

```diff
@@ -0,0 +1,123 @@
+  private _ReferenceERC20: ERC20
+
+  private static _address2ERC20cache: { [K: string]: ERC20 } = {}
+
+  public constructor(web3: Web3) {
+    this._ReferenceERC20 = new web3.eth.Contract(Erc20ABI)
+
+    // TODO remove later
+    ;(window as any).erc20 = this._ReferenceERC20
+  }
+
+  public async balanceOf(tokenAddress: string, userAddress: string): Promise<BN> {
+    console.log('userAddress', userAddress)
+    console.log('tokenAddress', tokenAddress)
+    if (!userAddress || !tokenAddress) return ZERO
+
+    const erc20 = this._getERC20AtAddress(tokenAddress)
+
+    const result = await erc20.methods.balanceOf(userAddress).call()
+    console.log('balance result for token', tokenAddress, 'user', userAddress, '=', result) // this is string
+
+    return toBN(result)
+  }
+
+  public async allowance(tokenAddress: string, userAddress: string, spenderAddress: string): Promise<BN> {
+    if (!userAddress || !tokenAddress) return ZERO
+
+    const erc20 = this._getERC20AtAddress(tokenAddress)
+
+    const result = await erc20.methods.allowance(userAddress, spenderAddress).call()
+    console.log('allowance result', result)
+
+    return toBN(result)
+  }
+
+  public async approve(
+    tokenAddress: string,
+    userAddress: string,
+    spenderAddress: string,
+    amount: BN,
+    txOptionalParams?: TxOptionalParams,
+  ): Promise<Receipt> {
+    const erc20 = this._getERC20AtAddress(tokenAddress)
+
+    const tx = erc20.methods.approve(spenderAddress, amount).send({
+      from: userAddress,
+    })
+
+    if (txOptionalParams && txOptionalParams.onSentTransaction) {
+      tx.once('receipt', txOptionalParams.onSentTransaction)
+    }
+
+    return tx
+  }
+
+  public async transfer(
+    tokenAddress: string,
+    fromAddress: string,
+    toAddress: string,
+    amount: BN,
+    txOptionalParams?: TxOptionalParams,
+  ): Promise<Receipt> {
+    const erc20 = this._getERC20AtAddress(tokenAddress)
+
+    const tx = erc20.methods.transfer(toAddress, amount).send({
+      from: fromAddress,
+    })
+
+    if (txOptionalParams) {
+      tx.once('receipt', txOptionalParams.onSentTransaction)
+    }
+
+    return tx
+  }
+
+  public async transferFrom(
+    senderAddress: string,
+    tokenAddress: string,
+    fromAddress: string,
+    toAddress: string,
+    amount: BN,
+    txOptionalParams?: TxOptionalParams,
+  ): Promise<Receipt> {
+    const erc20 = this._getERC20AtAddress(tokenAddress)
+
+    const tx = erc20.methods.transferFrom(senderAddress, toAddress, amount).send({
+      from: fromAddress,
+    })
+
+    if (txOptionalParams) {
+      tx.once('receipt', txOptionalParams.onSentTransaction)
+    }
+
+    return tx
+  }
+
+  /********************************    private methods   ********************************/
+
+  private _getERC20AtAddress(address: string): ERC20 {
+    const erc20 = Erc20ApiImpl._address2ERC20cache[address]
+
+    if (erc20) return erc20
+
+    const newERC20 = this._ReferenceERC20.clone()
```

Comment:
It more natural to not keep a reference implementation and in this line do a:
```js
const erc20Contract = new web3.eth.Contract(Erc20ABI, address)
this._contractsCache[address] = erc20Contract
```

nit: Regarding `_address2ERC20cache`. In the app `ERC20` when used in camel case, we use it like `erc20`, or `whateverErc20`. Anyways, `_contractsCache` seems like a simpler name

Reply:
`clone` retains defaults (which granted we don't use here), but also skips unnecessary stuff like abi decoding, etc. So it's a recommended way
Yes, it would be ideal if it accepted a new address (`clone(address)`), but it doesn't

Reply:
Will change to `_contractsCache`

Reply:
> clone retains defaults (which granted we don't use here), but also skips unnecessary stuff like abi decoding

Cool, didn't know!